src/graph.jl

```diff
@@ -115,10 +116,89 @@ function Base.show(io::IO, ::MIME"text/plain", d::Graph)
     end
 end
 
+"""
+    HeteroGraph(; kws...)
+
+HeteroGraph is used for HeteroGeneous Graphs.
+
+`HeteroGraph` unlike `Graph` can have different types of nodes. Each node pertains to different types of information. 
+
+Edges in `HeteroGraph` is defined by relations. A relation is a tuple of 
+(`src_node_type`, `edge_type`, `target_node_type`) where `edge_type` represents the relation
+between the src and target nodes. Edges between same node types are possible. 
+
+A `HeteroGraph` can be directed or undirected. It doesn't distinguish between directed 
+or undirected graphs. Therefore, for undirected graphs, it will store edges in both directions.
+Nodes are indexed in `1:num_nodes`.
+
+# Keyword Arguments
+
+- `num_nodes`: Dictionary containing the number of nodes for each node type. If omitted, is inferred from `edge_index`.
+- `num_edges`: Dictionary containing the number of edges for each relation.
+- `edge_indices`: Dictionary containing the `edge_index` for each edge relation. An `edge_index` is a tuple containing two vectors with length equal to the number of edges for the relation.
+    The first vector contains the list of the source nodes of each edge, the second contains the target nodes.
+- `node_data`: node-related data. Can be `nothing`, a named tuple of arrays or a dictionary of arrays.
+            The array's last dimension size should be equal to the number of nodes.
+            Default `nothing`.
+- `edge_data`: edge-related data. Can be `nothing`, a named tuple of arrays or a dictionary of arrays.
+            The array's last dimension size should be equal to the number of edges.
+            Default `nothing`.
+"""
+struct HeteroGraph <: AbstractGraph
+    num_nodes::Dict{String, Int}
+    num_edges::Dict{Tuple{String, String, String}, Int}
+    edge_indices::Dict{Tuple{String, String, String}, Tuple{Vector{Int}, Vector{Int}}}
+    node_data
+    edge_data
+end
+
+function HeteroGraph(;
+    num_nodes::Union{Dict{String, Int}, Nothing}=nothing,
```

Comment:
do we need `nothing`? can't we just use `Dict{String, Int}()` as default?

Reply:
Assigning `num_nodes` `Dict{String, Int}()` might indicate that there is no node in the Graph. Whereas `nothing` indicates that `num_nodes` information is not provided but can be inferred from other values.